src/BinWrapper.jl

```diff
@@ -0,0 +1,161 @@
+  if tempfile
+    # create temporary files
+    cmdin_p, cmdin    = mktemp()
+    cmdout_p, cmdout  = mktemp()
+    cmderr_p, cmderr  = mktemp()
+
+    # save input
+    write(cmdin, input)
+    close(cmdin)
+    cmdin = open(cmdin_p)
+
+    # run a process
+    process = run(pipeline(ignorestatus(cmd), stdin=cmdin, stdout=cmdout, stderr=cmderr))
+
+    # read out/err of a command
+    cmdout = open(cmdout_p)
+    cmderr = open(cmderr_p)
+
+    out = read(cmdout, String)
+    err = read(cmderr, String)
+
+    # return all the needed data
+    return (
+      out     = out,
+      err     = err,
+      ret     = process.exitcode
+    )
+  else # non-tempfile-based implementation
+    # create pipes/buffers for in/out/err of a command
+    cmdin  = IOBuffer(input)
+    cmdout = IOBuffer()
+    cmderr = IOBuffer()
+
+    if VERSION < v"1.1"
+      cmdin  = Pipe()
+      cmdout = Pipe()
+      cmderr = Pipe()
+    end
+
+    # run a process
+    process = run(pipeline(ignorestatus(cmd), stdin=cmdin, stdout=cmdout, stderr=cmderr), wait=!(VERSION < v"1.1"))
+
+    if VERSION < v"1.1"
+      # close unnecessary communication tunnels of pipes initialized by pipeline()
+      close(cmdout.in)
+      close(cmderr.in)
+
+      # pass the input to the process and close the related pipe
+      # note: this could be also `write(process, input)`, but then `cmdin` would not be used which is not cool
+      write(cmdin, input)
+      close(cmdin)
+
+      # asynchronously read the outputs of the command
+      out = @async String(read(cmdout))
+      err = @async String(read(cmderr))
+
+      # return all the needed data
+      return (
+        out     = fetch(out),
+        err     = fetch(err),
+        ret     = process.exitcode
+      )
+    end # if VERSION < v"1.1"
+
+    # read the outputs of the command
+    out = String(take!(cmdout))
+    err = String(take!(cmderr))
+
+    # return all the needed data
+    return (
+      out     = out,
+      err     = err,
+      ret     = process.exitcode
+    )
+  end
+end
+
+# This function wraps the binaries that come with the original kbmag library.
+# Here, all the parameters are passed as Strings
+function kbmag_bin_wrapper(program::String; params = (), input::String = "", force_dir::String = ".")
+  # remember the current working directory
+  call_dir = pwd()
+
+  # change the directory if necessary
+  # WARNING: when calling a kbmag binary we need to make sure that
+  # all the input files are in the current working directory
+  if force_dir != "."
+    cd(force_dir)
+  end
+
+  result = low_level_exec(`$kbmag_bin_dir$program $(params)`; input = input)
+
+  # change the current working directory back to the previous one
+  if force_dir != "."
+    cd(call_dir)
+  end
+
+  return result # (out, err, ret)
+end
+
+# This function wraps a kbprog binary
+#     groupname::String   --- input file name
+#     dir::String         --- working directory
+function kbprog_call(groupname::String;
```

Comment:
The reason to keep it apart is that the `wordreduce` program uses a number of files that have the same basename but different extensions. If we require the `groupname` parameter to be a path to a file, it could be misleading. How the user could possibly know that it is not a path to a single file but to a number of different files? If we keep the groupname and the execution path separately, it is at least less misleading. What do you think about that?

Reply:
And of course `execution_dir` can point anywhere.

Reply:
But from what I see even call to `wordreduce` needs the RWS-describing file anyway?
then just give it the file.

I'm not against keeping this as is as long as it is not user-facing, I'm for adding a user-facing function with the described semantics.

Reply:
`wordreduce` needs a number of files with the same basename. For `237` it would be `237`, `237.kbprog` and `237.reduce` (the `237.kbprog.ec` file is not required actually). The idea behind such a construction is that you can do the following (minimal example):
```
$ cd some_directory
$ ls
group
$ julia
julia> using KBmag
julia> KBmag.wordreduce_call("group", ["word"]) # no need to pass execution_dir
1-element Array{Any,1}:
 "result"
```

Reply:
`wordreduce` depends only on the file describing the RWS and generates the rest on demand.

```julia
function kbprog_call(rws_file::AbstractString)
    rws_file = abspath(rws_file)
    isfile(rws_file) || error(...)
    groupfile, exec_dir = basename(rws_file), dirname(rws_file)
    return kbprog_call(groupfile, exec_dir)
end

function wordreduce_call(rws_file::AbstractString, words::AbstractVector{<:AbstractString})
    rws_file = abspath(rws_file)
    isfile(rws_file) || error(...)
    groupfile, exec_dir = basename(rws_file), dirname(rws_file)
    return kbprog_call(groupfile, exec_dir, String[words...])
end
```
both `KBmag.wordreduce_call("group", ["word"])` and `KBmag.wordreduce_call("groups/237", ["word"])` work just fine (and are tab completed).

Reply:
That is only half true. The `wordreduce` binary *does* call proper `.kbprog` and `.reduce` files, based on the groupname. So its successful execution depends on _two_ files with the same basename, but different extensions.

It seems like we have some kind of a clash here. You want it to be more user-friendly which I fully understand. From this point of view, another parameter --- `execution_dir` --- is unnecessary. On the other hand, I want it to be separated as `kbprog_call` generates new files in a base directory and `wordreduce_call` works on two different files. This way keeping a basename apart seems to be a foolproof approach.

I know that you will be the first one to use this package so I have loved to make it more handy for you :)